components/engine_test/src/lib.rs

```diff
@@ -243,60 +247,80 @@ pub mod kv {
-        }
+            let mut reg = self.registry.lock().unwrap();
+            if let Some(suffix) = suffix {
+                if let Some(db) = reg.get(&(id, suffix)) {
+                    // Target tablet exist in the cache
+
+                    if options.create_new() {
+                        return Err(box_err!(
+                            "region {} {} already exists",
+                            id,
+                            db.as_inner().path()
+                        ));
+                    }
+                    return Ok(db.clone());
+                } else if !options.cache_only() {
+                    let tablet_path = self.tablet_path(id, suffix);
+                    // Even though neither options.create nor options.create_new are true, if the
+                    // tablet files already exists, we will open it by calling
+                    // inner.create_tablet. In this case, the tablet exists but not in the cache
+                    // (registry).
+                    if options.create()
+                        || options.create_new()
+                        || KvTestEngine::exists(tablet_path.to_str().unwrap_or_default())
+                    {
+                        let kv_engine = self.inner.create_tablet(&tablet_path)?;
+                        if !options.skip_cache() {
+                            reg.insert((id, suffix), kv_engine.clone());
+                        }
+                        return Ok(kv_engine);
+                    }
+                }
+            } else if options.cache_only() {
+                // This branch reads an arbitrary tablet with region id `id`
 
-        fn open_tablet(&self, id: u64, suffix: u64) -> Result<KvTestEngine> {
-            let reg = self.registry.lock().unwrap();
-            if let Some(db) = reg.get(&(id, suffix)) {
-                return Ok(db.clone());
+                if let Some(k) = reg.keys().find(|k| k.0 == id) {
+                    return Ok(reg.get(k).unwrap().clone());
+                }
             }
 
-            let db_path = self.tablet_path(id, suffix);
-            let db = self.open_tablet_raw(db_path.as_path(), false)?;
-            Ok(db)
-        }
-
-        fn open_tablet_cache(&self, id: u64, suffix: u64) -> Option<KvTestEngine> {
-            self.registry.lock().unwrap().get(&(id, suffix)).cloned()
+            Err(box_err!(
+                "tablet with region id {} suffix {:?} does not exist",
+                id,
+                suffix
+            ))
         }
 
-        fn open_tablet_cache_any(&self, id: u64) -> Option<KvTestEngine> {
-            let reg = self.registry.lock().unwrap();
-            if let Some(k) = reg.keys().find(|k| k.0 == id) {
-                return Some(reg.get(k).unwrap().clone());
-            }
-            None
-        }
-
-        fn open_tablet_raw(&self, path: &Path, _readonly: bool) -> Result<KvTestEngine> {
+        fn open_tablet_raw(&self, path: &Path) -> Result<KvTestEngine> {
```

Comment:
The path may not be named correctly. For example, it may be just a temporary path.

Reply:
Good catch.